subprojects/core/src/main/java/org/gradle/execution/plan/TaskNodeFactory.java

```diff
@@ -64,4 +82,55 @@ public void clear() {
         nodes.clear();
     }
 
+    private static class DefaultTypeOriginInspectorFactory {
+        private final Map<Project, ProjectScopedTypeOriginInspector> projectToInspector = Maps.newConcurrentMap();
+        private final Map<Class<?>, File> clazzToFile = Maps.newConcurrentMap();
+
+        public ProjectScopedTypeOriginInspector forTask(Task task) {
+            return projectToInspector.computeIfAbsent(task.getProject(), ProjectScopedTypeOriginInspector::new);
+        }
+
+        @Nullable
+        private File jarFileFor(Class<?> pluginClass) {
+            return clazzToFile.computeIfAbsent(pluginClass, clazz -> toFile(pluginClass.getProtectionDomain().getCodeSource().getLocation()));
+        }
+
+        @Nullable
+        private static File toFile(URL url) {
+            try {
+                return new File(url.toURI());
+            } catch (URISyntaxException e) {
+                return null;
+            }
+        }
+
+        private class ProjectScopedTypeOriginInspector implements WorkValidationContext.TypeOriginInspector {
+            private final PluginContainer plugins;
+            private final PluginManagerInternal pluginManager;
+            private final Map<Class<?>, Optional<PluginId>> clazzToPlugin = Maps.newConcurrentMap();
```

Comment:
```suggestion
            private final Map<Class<?>, Optional<PluginId>> classToPlugin = Maps.newConcurrentMap();
```

Nit: No reason to use `clazz` here.

Reply:
fixed